File: examples/perl/subscribe_async.pl
#!/usr/bin/env perl

# Copyright 2007-2022 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.

#
# @file subscribe_async.pl
#
# @brief This example program provides a basic demonstration of subscribing
# to receive a message from the GMSEC Bus using a callback.
#


use strict;
use libgmsec_perl;

*isa = \&UNIVERSAL::isa;


# Define custom callback to receive messages
#
{
    package MyCallback;
    use base 'libgmsec_perl::Callback';

    sub new
    {
        my $class = shift;
        my $self = $class->SUPER::new();
        bless $self, $class;
        return $self;
    }

    sub onMessage
    {
        my ($self, $connection, $msg) = @_;
        libgmsec_perl::logInfo("MyCallback.onMessage -- \n" . $msg->toXML());
    }
}


# Define custom callback to receive events
# 
# Although not required, an event callback can assist with determining
# whether we have a connection to the GMSEC Bus, or in the case of this
# example program, whether the Auto-Dispatcher thread has terminated
# due to some error.
# 
{
    package MyEventCallback;
    use base 'libgmsec_perl::EventCallback';

    sub new
    {
        my $class = shift;
        my $self = $class->SUPER::new();
        bless $self, $class;

        $self->{ERROR} = 0;

        return $self;
    }

    sub onEvent
    {
        my ($self, $connection, $status, $event) = @_;

        # Check event type and/or error status of the Status object.
        if ($event == $libgmsec_perl::Connection::Event_DISPATCHER_ERROR_EVENT)
        {
            # Auto-dispatcher error has occurred we are done!
            libgmsec_perl::logError("MyEventCallback.onEvent -- " . $status->getReason());
            $self->{ERROR} = 1;
        }
        elsif ($status->hasError())
        {
            # Soft-error has occurred
            libgmsec_perl::logWarning("MyEventCallback.onEvent -- " . $status->getReason());
        }
        else
        {
            # Nominal event has occurred
            libgmsec_perl::logInfo("MyEventCallback.onEvent -- " . $status->getReason());
        }
    }

    sub haveError
    {
        my ($self) = @_;

        return $self->{ERROR};
    }
}


sub main
{
    # Load the command-line input into a GMSEC Config object
    my $config = undef;
    if ($#ARGV ge 0)
    {
        $config = libgmsec_perl::Config::create(\@ARGV);
    }
    else
    {
        $config = libgmsec_perl::Config->new();
    }

    # Initialize log level for output
    initializeLogLevel($config);

    # Declared custom callback and custom event callback
    my $mcb = MyCallback->new();
    my $ecb = MyEventCallback->new();

    # Deduce subscription topic for the example program
    my $subscriptionTopic = "";
    my $level = $config->getIntegerValue("gmsec-schema-level", $libgmsec_perl::Specification::SchemaLevel_LEVEL_0);

    if ($level == $libgmsec_perl::Specification::SchemaLevel_LEVEL_0)
    {
        $subscriptionTopic = "C2MS.>";
    }
    elsif ($level == $libgmsec_perl::Specification::SchemaLevel_LEVEL_1)
    {
        $subscriptionTopic = "C2MS-PIPE.>";
    }
    else
    {
        $subscriptionTopic = "GMSEC.>";
    }

    eval
    {
        # Create connection instance.
        my $conn = libgmsec_perl::Connection::create( $config );

        # Register custom EventCallback
        $conn->registerEventCallback( $libgmsec_perl::Connection::Event_ALL_EVENTS, $ecb );

        # Establish connection to the GMSEC Bus.
        $conn->connect();

        # Output information regarding the API version and connection
        libgmsec_perl::logInfo(libgmsec_perl::Connection::getAPIVersion());
        libgmsec_perl::logInfo("Middleware version = " . $conn->getLibraryVersion());

        # Subscribe to receive message via a callback
        $conn->subscribe( $subscriptionTopic, $mcb );
        libgmsec_perl::logInfo("Subscribed to topic: " . $subscriptionTopic);

        # Attempt to receive a message
        my $rcvd = $conn->receive( $libgmsec_perl::GMSEC_WAIT_FOREVER );

        # It is good practice to check if a message was received, but in this particular
        # case, since GMSEC_WAIT_FOREVER is being used, we are either guaranteed to
        # receive a message or have some exception thrown due to a middleware issue.
        if (defined $rcvd)
        {
            libgmsec_perl::logInfo("Received a message; dispatching it...");
            $conn->dispatch( $rcvd );
            $rcvd->acknowledge();
            libgmsec_perl::Message::destroy( $rcvd );
        }
        else
        {
            libgmsec_perl::logWarning("Failed to receive a message");
        }

        # Disconnect from the GMSEC Bus, and terminate subscription.
        $conn->disconnect();

        # Destroy the connection instance
        libgmsec_perl::Connection::destroy($conn);
    };
    if (isa($@, 'libgmsec_perl::GmsecException'))
    {
        # If here, the most likely case is the middleware could not be
        # deduced, or a failure occurred when attempting to connect to
        # the GMSEC Bus.
        my $error = $@;
        libgmsec_perl::logError("Exception: " . $error->what());
        return -1;
    }

    return 0;
}


sub initializeLogLevel
{
    my ($config) = @_;
    my $level = libgmsec_perl::Log::fromString( $config->getValue("loglevel", "info") );
    libgmsec_perl::Log::setReportingLevel($level);
}


# Main entry point
#
exit( main() );

File: examples/ruby/subscribe_async.rb
#!/usr/bin/env ruby

# Copyright 2007-2023 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.

#
# @file subscribe_async.rb
#
# @brief This example program provides a basic demonstration of subscribing
# to receive a message from the GMSEC Bus using a callback.
#


require 'libgmsec_ruby'
require './logger'


# Define custom callback to receive messages
#
class MyCallback < Libgmsec_ruby::Callback
    def initialize()
        super()
    end

    def on_message(conn, msg)
        Libgmsec_ruby::Log::info("MyCallback.on_message -- \n" + msg.to_xml())
        msg.acknowledge()
    end
end


# Define custom callback to receive events
# 
# Although not required, an event callback can assist with determining
# whether we have a connection to the GMSEC Bus, or in the case of this
# example program, whether the Auto-Dispatcher thread has terminated
# due to some error.
# 
class MyEventCallback < Libgmsec_ruby::EventCallback
    def initialize()
        super()
        @error = false
    end

    def on_event(conn, status, event)
        # Check event type and/or error status of the Status object.

        if event == Libgmsec_ruby::Connection::Event_DISPATCHER_ERROR_EVENT
            # Auto-dispatcher error has occurred we are done!
            Libgmsec_ruby::Log::error("MyEventCallback.on_event -- " + status.get_reason())
            self.error = True

        elsif status.has_error()
            # Soft-error has occurred
            Libgmsec_ruby::Log::warning("MyEventCallback.on_event -- " + status.get_reason())

        else
            # Nominal event has occurred
            Libgmsec_ruby::Log::info("MyEventCallback.on_event -- " + status.get_reason())
        end
    end

    def have_error()
        @error
    end
end


class Subscribe_Async
    def initialize(args)
        # Set up connection configuration options using values from the command line.
        @config = Libgmsec_ruby::Config.new( args )

        # Initialize log level for output
        initialize_log_level()

        # Initialize log handler
        @logHandler = CustomLogHandler.new
        Libgmsec_ruby::Log.register_handler(@logHandler)

        # Declared custom callback and custom event callback
        @mcb = MyCallback.new()
        @ecb = MyEventCallback.new()

        # Deduce subscription topic for the example program
        level = @config.get_integer_value("gmsec-schema-level", Libgmsec_ruby::Specification::SchemaLevel_LEVEL_0)

        case level
        when Libgmsec_ruby::Specification::SchemaLevel_LEVEL_0
            @subscription_topic = "C2MS.>"
        when Libgmsec_ruby::Specification::SchemaLevel_LEVEL_1
            @subscription_topic = "C2MS-PIPE.>"
        else
            @subscription_topic = "GMSEC.>"
        end
    end


    def initialize_log_level()
        level = Libgmsec_ruby::Log::from_string( @config.get_value("loglevel", "info") )
        Libgmsec_ruby::Log::set_reporting_level(level)
    end


    def run()
        # Create connection instance.
        conn = Libgmsec_ruby::Connection::create( @config )

        # Register custom EventCallback
        conn.register_event_callback(Libgmsec_ruby::Connection::Event_ALL_EVENTS, @ecb)

        # Establish connection to the GMSEC Bus.
        conn.connect()

        # Output information regarding the API version and connection
        Libgmsec_ruby::Log::info(Libgmsec_ruby::Connection::get_api_version())
        Libgmsec_ruby::Log::info("Middleware version = " + conn.get_library_version())

        # Subscribe to receive message via a callback
        conn.subscribe(@subscription_topic, @mcb)
        Libgmsec_ruby::Log::info("Subscribed to topic: " + @subscription_topic)

        # Attempt to receive a message
        rcvd = conn.receive(Libgmsec_ruby::GMSEC_WAIT_FOREVER)

        # It is good practice to check if a message was received, but in this particular
        # case, since GMSEC_WAIT_FOREVER is being used, we are either guaranteed to
        # receive a message or have some exception thrown due to a middleware issue.
        if rcvd == nil
            Libgmsec_ruby::Log::warning("Failed to receive a message")
        else
            Libgmsec_ruby::Log::info("Received a message; dispatching it...")
            conn.dispatch(rcvd)
            rcvd.acknowledge()
            Libgmsec_ruby::Message::destroy(rcvd)
        end

        # Disconnect from the GMSEC Bus, and terminate subscription.
        conn.disconnect()

        # Destroy the connection instance
        Libgmsec_ruby::Connection::destroy(conn)
    end
end


# Main entry point
begin
    ex = Subscribe_Async.new( ARGV )
    ex.run
rescue GmsecException => e
    # If here, the most likely case is the middleware could not be
    # deduced, or a failure occurred when attempting to connect to
    # the GMSEC Bus.
    Libgmsec_ruby::Log::error("Exception: #{e.message}")
end
